crates/net/src/eventsource/futures.rs

```diff
@@ -0,0 +1,258 @@
+//! A wrapper around the `EventSource` API using the Futures API to be used with async rust.
+//!
+//! EventSource is similar to WebSocket with the major differences being:
+//!
+//! * they are a one-way stream of server generated events
+//! * their connection is managed entirely by the browser
+//! * their data is slightly more structured including an id, type and data
+//!
+//! EventSource is therefore suitable for simpler scenarios than WebSocket.
+//!
+//! See the [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) to learn more.
+//!
+//! # Example
+//!
+//! ```rust
+//! use gloo_net::eventsource::futures::EventSource;
+//! use wasm_bindgen_futures::spawn_local;
+//! use futures::StreamExt;
+//!
+//! # macro_rules! console_log {
+//! #    ($($expr:expr),*) => {{}};
+//! # }
+//! # fn no_run() {
+//! let mut es = EventSource::new("http://api.example.com/ssedemo.php").unwrap();
+//! es.subscribe_event("some-event-type").unwrap();
+//! es.subscribe_event("another-event-type").unwrap();
+//!
+//! spawn_local(async move {
+//!     while let Some(Ok((event_type, msg))) = es.next().await {
+//!         console_log!(format!("1. {}: {:?}", event_type, msg))
+//!     }
+//!     console_log!("EventSource Closed");
+//! })
+//! # }
+//! ```
+use crate::eventsource::{EventSourceError, State};
+use crate::js_to_js_error;
+use futures_channel::mpsc;
+use futures_core::{ready, Stream};
+use gloo_utils::errors::JsError;
+use pin_project::{pin_project, pinned_drop};
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::ops::DerefMut;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::MessageEvent;
+
+/// Wrapper around browser's EventSource API.
+#[allow(missing_debug_implementations)]
+#[pin_project(PinnedDrop)]
+pub struct EventSource {
+    es: web_sys::EventSource,
+    message_sender: mpsc::UnboundedSender<StreamMessage>,
+    #[pin]
+    message_receiver: mpsc::UnboundedReceiver<StreamMessage>,
+    #[allow(clippy::type_complexity)]
+    closures: Arc<
+        Mutex<(
+            HashMap<String, Closure<dyn FnMut(MessageEvent)>>,
+            Closure<dyn FnMut(web_sys::Event)>,
+        )>,
+    >,
+}
+
+impl EventSource {
+    /// Establish an EventSource.
+    ///
+    /// This function may error in the following cases:
+    /// - The connection url is invalid
+    ///
+    /// The error returned is [`JsError`]. See the
+    /// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventSource/EventSource#exceptions_thrown)
+    /// to learn more.
+    pub fn new(url: &str) -> Result<Self, JsError> {
```

Comment:
I think we should also have a function `new_with_subscriptions` which calls new and subscribes to the events passed in

It can made more efficient as there's no need to create an empty HashMap and then mutate it later on.

Reply:
I'm unsure about this as we'd still be mutating a hashmap in the constructor... The downside of doing this would also be more code in the implementation.

Reply:
Not sure if this efficiency gain is worth the additional complexity. Personally I am happy with creating an EventSource and then subscribing to it.

Reply:
There's `HashMap::with_capacity` but it's fine to leave it as it is as well